projects/vdk-plugins/vdk-meta-jobs/src/vdk/plugin/meta_jobs/cached_data_job_executor.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import json
+import logging
+import os
+import time
+from typing import Dict
+
+from taurus_datajob_api import ApiException
+from vdk.internal.core.errors import ErrorMessage
+from vdk.internal.core.errors import UserCodeError
+from vdk.plugin.meta_jobs.meta import IDataJobExecutor
+from vdk.plugin.meta_jobs.meta import TrackableJob
+from vdk.plugin.meta_jobs.remote_data_job import JobStatus
+
+log = logging.getLogger(__name__)
+
+ACTIVE_JOB_STATUSES = [JobStatus.SUBMITTED.value, JobStatus.RUNNING.value]
+
+
+class TrackingDataJobExecutor:
+    def __init__(self, executor: IDataJobExecutor):
+        self._executor = executor
+        self._jobs_cache: Dict[str, TrackableJob] = dict()
+        self._time_between_status_check_seconds = int(
+            os.environ.get("VDK_META_JOBS_TIME_BETWEEN_STATUS_CHECK_SECONDS", "40")
+        )
+
+    def register_job(self, job: TrackableJob):
+        if job.job_name in self._jobs_cache:
+            log.warning(
+                f"Job with name {job.job_name} aleady exists. Details: {self._jobs_cache[job.job_name]}. "
+                f"Will overwrite it."
+            )
+        self._jobs_cache[job.job_name] = job
+
+    def start_job(self, job_name: str) -> None:
+        """
+        :param job_name: the job to start and track
+        """
+        job = self.__get_job(job_name)
+        job.start_attempt += 1
+        execution_id = self._executor.start_job(job.job_name, job.team_name)
+        log.info(f"Starting new data job execution with id {execution_id}")
+        job.execution_id = execution_id
+        job.status = JobStatus.SUBMITTED.value
+        job.details = self._executor.details_job(
+            job.job_name, job.team_name, job.execution_id
+        )
+        log.info(
+            f"Started data job {job_name}:\n{self.__get_printable_details(job.details)}"
+        )
+
+    def finalize_job(self, job_name):
+        job = self.__get_job(job_name)
+        details = self._executor.details_job(
+            job.job_name, job.team_name, job.execution_id
+        )
+        job.details = details
+        log.info(
+            f"Finished data job {job_name}:\n"
+            f"start_time: {details['start_time']}\n"
+            f"end_time: {details['end_time']}\n"
+            f"status: {details['status']}\n"
+            f"message: {details['message']}"
+        )
+        if job.status != JobStatus.SUCCEEDED.value and job.fail_meta_job_on_error:
+            raise UserCodeError(
```

Comment:
If the job failed with a platform error, should this raise a PlatformError exception instead?

Reply:
We discussed this during the hackaton with our users (who also participated) and we had 2 reasons to do it this way: 

The thought is that if we raise a platform error, this will trigger automatic re-try of the meta job. 
If the orechestated job has 4 re-tries then meta job has 4 re-tries this is 16 re-tries of orchestrated job. 

I cannot recall the second reason now. That's why one should write comments ...